// build-fail
// build-fail
// normalize-stderr-test: "c\[.*?\]" -> "c[HASH]"
#![allow(incomplete_features)]
#![feature(adt_const_params, rustc_attrs)]

pub struct Str<const S: &'static str>;

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling-alt(<c::Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ">>)
//~| ERROR demangling-alt(<c::Str<"abc">>)
impl Str<"âˆ‚Ã¼"> {}

#[rustc_symbol_name]
//~| ERROR demangling-alt(<c::Str<"áƒ¡áƒáƒ­áƒ›áƒ”áƒšáƒáƒ“_áƒ’áƒ”áƒ›áƒ áƒ˜áƒ”áƒšáƒ˜_áƒ¡áƒáƒ“áƒ˜áƒšáƒ˜">>)
//~| ERROR demangling
//~| ERROR demangling
impl Str<"áƒ¡áƒáƒ­áƒ›áƒ”áƒšáƒáƒ“_áƒ’áƒ”áƒ›áƒ áƒ˜áƒ”áƒšáƒ˜_áƒ¡áƒáƒ“áƒ˜áƒšáƒ˜"> {}

#[rustc_symbol_name]
//~| ERROR demangling-alt(<c::Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ">>)
//~| ERROR demangling-alt(<c::Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ">>)
//~| ERROR demangling-alt(<c::Str<"\t\n">>)
impl Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ"> {}

#[feature(adt_const_params, rustc_attrs)]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"âˆ‚Ã¼">>)
impl Str<"âˆ‚Ã¼"> {}

#[allow(incomplete_features)]
//~| ERROR demangling-alt(<c::Str<"âˆ‚Ã¼">>)
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"áƒ¡áƒáƒ­áƒ›áƒ”áƒšáƒáƒ“_áƒ’áƒ”áƒ›áƒ áƒ˜áƒ”áƒšáƒ˜_áƒ¡áƒáƒ“áƒ˜áƒšáƒ˜">>)
impl Str<"abc"> {}

#[allow(incomplete_features)]
//~| ERROR demangling-alt(<c::Str<"abc">>)
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ">>)
impl Str<"'"> {}

fn main() {}
