// build-fail
// compile-flags: -C symbol-mangling-version=v0 --crate-name=c
//~| ERROR demangling-alt(<c::Str<"\t\n">>)
#![feature(adt_const_params, rustc_attrs)]
#![rustc_symbol_name]

pub struct Str<const rustc_symbol_name: &'static str>;

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"abc">>)
impl Str<"abc"> {}

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"'">>)
impl Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ"> {}

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"\t\n">>)
impl Str<"\t\n"> {}

#[allow(incomplete_features)]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"âˆ‚Ã¼">>)
impl Str<"âˆ‚Ã¼"> {}

#[rustc_symbol_name]
//~^ ERROR symbol-name
// compile-flags: -C symbol-mangling-version=v0 --crate-name=c
//~| ERROR demangling
impl Str<"áƒ¡áƒáƒ­áƒ›áƒ”áƒšáƒáƒ“_áƒ’áƒ”áƒ›áƒ áƒ˜áƒ”áƒšáƒ˜_áƒ¡áƒáƒ“áƒ˜áƒšáƒ˜"> {}

#[allow(incomplete_features)]
//~^ ERROR symbol-name
// compile-flags: -C symbol-mangling-version=v0 --crate-name=c
//~| ERROR demangling-alt(<c::Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ">>)
impl Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ"> {}

fn main() {}
