// build-fail
// compile-flags: -C symbol-mangling-version=v0 --crate-name=c
// normalize-stderr-test: "c\[.*?\]" -> "c[HASH]"
#![allow(incomplete_features)]
#![feature(adt_const_params, rustc_attrs)]

pub struct Str<const S: &'static str>;

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"abc">>)
impl Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ"> {}

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling-alt(<c::Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ">>)
//~| ERROR demangling-alt(<c::Str<"'">>)
impl Str<"áƒ¡áƒáƒ­áƒ›áƒ”áƒšáƒáƒ“_áƒ’áƒ”áƒ›áƒ áƒ˜áƒ”áƒšáƒ˜_áƒ¡áƒáƒ“áƒ˜áƒšáƒ˜"> {}

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"\t\n">>)
impl Str<"abc"> {}

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling-alt(<c::Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ">>)
//~| ERROR demangling-alt(<c::Str<"'">>)
impl Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ"> {}

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"áƒ¡áƒáƒ­áƒ›áƒ”áƒšáƒáƒ“_áƒ’áƒ”áƒ›áƒ áƒ˜áƒ”áƒšáƒ˜_áƒ¡áƒáƒ“áƒ˜áƒšáƒ˜">>)
impl Str<"'"> {}

#[rustc_symbol_name]
//~^ ERROR symbol-name
//~| ERROR demangling
//~| ERROR demangling-alt(<c::Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ">>)
impl Str<"ðŸŠðŸ¦ˆðŸ¦†ðŸ® Â§ ðŸ¶ðŸ‘’â˜•ðŸ”¥ Â§ ðŸ§¡ðŸ’›ðŸ’šðŸ’™ðŸ’œ"> {}

fn main() {}
