fn main() {
    ...........................................................................................................................................................................................................................({.............................................................................................................................................................................................},)

    fn main() {
   
 ...............................................................[......................................................................................................................................................................................................................................................................................................................................................................................................[.................................................................................................................................................................................................................., ......................................................... async {.................................................................................................................................................................................}], ...................................................................................................................................................................................................................................................................................................[....................................................................................................................................................................................................................................................................................................................................[.................................................................................................................................................................................................................., ......................................................... async {...............................................................................................................................................................................................................................................................................................................[.................................................................................................................................................................................................................., ......................................................... async {.................................................................................................................................................................................}]}], ........................................................................................................................................................................................................................................................................................................................................................................[.................................................................................................................................................................................................................., ......................................................... async {.................................................................................................................................................................................}]]]   //~^ ERROR E0005
}
    //~^ ERROR E0005
}
