fn main() {
    .......................................................................................(0)

    fn main() {
   
 ............................................................................................................................................................[...........................(......................................................................................................................................................................................................(...................................................................................................[...........................(......................................................................................................................................................................................................(.................................................................................................................................................................................................................................................................................................(0),),), .........................................................................................................................................................................................................................................................].(0),),), .........................................................................................................................................................................................................................................................]   //~^ ERROR E0010
}
    //~^ ERROR E0005
}
