fn main() {
    ...........................................................................................................................................................................................................................({.............................................................................................................................................................................................},)

    fn main() {
   
 ...............................................................[................................................................................................................................................................................................................................................[.........................................................................................................................................................................................................................................., ..............................................................................................................................................................................................................................................................................................................................[....................................................................................................................................................................................................................................................................................................................................[.........................................................................................................................................................................................................................................., ..........................................................................................................................................................................................................................................], ..........................................................................................................................................................................................................................................]], ...............................................................................................................................................................................................................]   //~^ ERROR E0005
}
    //~^ ERROR E0005
}
