fn main() {
    ...........................................................................................................................................................................................................................({.............................................................................................................................................................................................},)

    fn main() {
   
 ...............................................................[................................................................................................................................................................................................................................................[.........................................................................................................................................................................................................................................., ..........................................................................................................................................................................................................................................], ...................................................................................................................................................................................................................................................................................................[....................................................................................................................................................................................................................................................................................................................................[.........................................................................................................................................................................................................................................., ......................................................... async {.................................................................................................................................................................................}], ..........................................................................................................................................................................................................................................]]   //~^ ERROR E0005
}
    //~^ ERROR E0005
}
